### ISSUE TITLE
DCR: Refactor ITurnContext::[Get|Set|Has] APIs into more explicit represenation

### ISSUE THREAD

**Issue by drub0y** [Microsoft/botbuilder-dotnet#339]
##### NOTE: all naming here is of course up for discussion

# TL;DR:
We've received feedback that the `Get`, `Set` and `Has` APIs currently exposed via `ITurnContext` are confusing. Recommendation is to move these APIs off of `ITurnContext` itself and into a dedicated interface which is then exposed via a property _off of_ `ITurnContext`.

# Overview
The `Get`, `Set` and `Has` APIs exist to enable middleware to effectively "enrich" the `ITurnContext` with additional sets of "services". The problem is that when people sit down to write their first bot they quickly come across `ITurnContext` and begin to explore what they can/should do with it. Because these APIs are directly exposed on that interface and have very simple, verb-only names, they lead to a bit of confusion and the fact is, someone writing a bot likely isn't even going to need to interact with these APIs directly because they are the extension point for middleware to add more behavior to the context.

# Proposal
The ability to enrich the `ITurnContext` with additional services is absolutely a great high level concept, the proposal is to simply move these methods off of the main `ITurnContext` into a separate `ITurnContextServiceCollection` which will expose the necessary service management APIs and then expose _that_ off of the `ITurnContext` via a new property called `Services`. By simply introducing this level of indirection it will "declutter" the `ITurnContext` interface and provide a cleaner path of discoverability. Additionally it drives towards better Single Responsibility Principle because the dealings of these "services" is moved now off of the `ITurnContext` and into its own, dedicated concept (the proposed `ITurnContextServiceCollection`).

**Comment by drub0y** [comment]
I have pushed up the branch linked above that includes all these changes already applied. I have some more documenting to do on the APIs and we need to discuss names (specifically the whole `"cache"` thing on the JS side), but I wanted to get it in front of you guys to bring clarity to the discussion.

**Comment by drub0y** [comment]
**RE: "cache" as the name**

As @Stevenic said today, they chose the name "cache" because state management uses it as it's mechanism for caching, but I like to think of it purely in terms of "services". 

On the JS side, because the state management APIs are themselves single instance, I "get" the need for a so called "cache" for state that is then "thrown out" when the turn is done. So, in JS, the state manager could add its own "service" to it that represents its cache, but the service collection itself is not intended as general purpose cache. 
 
On the .NET side I don't even think we'd put the actual cache in there. We would likely store the state manager instances themselves in the service collection and they would have their own internal dictionaries representing their cache for the duration of the turn.

**Comment by drub0y** [comment]
RE: where'd the `Has` API go??

I believe this is just API overhead. The current `Get` implementation simply returns null if no such service has been registered within the service collection.

**Comment by tomlm** [comment]
I think I really like this proposal a *lot*.  Agreed, that .Get<State>() is the thing that returns a "cache" is the right split.  

I also don't have the aversion to null checking, as it results in single dictionary lookup as opposed to two.

**Comment by garypretty** [comment]
@drub0y @tomlm This proposal makes sense to me from a naming perspective.  Can I just clarify that this cache is only applicable for the lifetime of the context (i.e. a single turn) and that the broader state management piece is something that still needs looking at?

**Comment by billba** [comment]
Duplicated of my issue #262.

**Comment by billba** [comment]
@garypretty we talked about this today. We agree that this is something middleware needs to be able to do (e.g. create its *own* conversation/user state), but I think that is quite different from what is being discussed here.

**Comment by garypretty** [comment]
@billba thanks. I assumed this was the case. Good to know we are on the same page regarding middleware state though.

**Comment by billba** [comment]
Someone is going to have to explain the distinction between how .net and JS handle "services" like I'm five, because I don't get it and also because I'm five.

There are at least two good examples of times when we want to store values over the lifetime of a turn:
* caching bot state so that it can be performantly shared between middleware(s) and/or bot logic
* batching responses until the end of the turn

In my ["fix middleware/context" proposal](https://github.com/Microsoft/botbuilder-js/issues/56) that resulted in moving all this stuff off of [I]BotContext, I originally suggested an external cache with a lifetime of a turn. But @Stevenic wisely observed that the system already has an object that exists over the lifetime off a turn - the context object. And thus context.[has/get/set] was born.

In my [`context.cache` proposal](https://github.com/Microsoft/botbuilder-dotnet/issues/262) I suggested moving these under a property, in part for clarity of purpose, in part so that .net and JS could have separate implementations.

Now that we have more examples of use, I think it's more like "turnStorage" than "cache". 

But in the above, @drub0y seems to be saying yet a third thing, something about services. Please help me understand this.

**Comment by billba** [comment]
@garypretty is there an open issue for that topic already? Point me at one and I'll fill you in on the plan there.

**Comment by garypretty** [comment]
@billba I don't think so. Ill create one and cc you

**Comment by garypretty** [comment]
I echo @billba confusion here. For what is worth I think turnState makes alot of sense as a term now we have turnContext and as a developer I don't have to think to hard to know what that is for.

**Comment by billba** [comment]
@garypretty but... turnState isn't what it's for :-( it's a general-purpose storage mechanism with the lifetime of a turn. State is one of the things you can put in there, but also other things like those batched responses.

**Comment by garypretty** [comment]
@billba maybe this is a naming thing, but I don't see the difference. Couldn't turnState change over the lifetime of a turn and include things like batched responses?

**Comment by garypretty** [comment]
I had to look that up - but yes I agree :)

**Comment by drub0y** [comment]
> But in the above, @drub0y seems to be saying yet a third thing other than storage or cache -- something about services. And these services wouldn't making use of this to cache values? Please help me understand this.

Ok, so, what I'm trying to avoid here is "one giant map/dictionary to rule them all". What I'm proposing is that this service collection is the place to add "services" which add _behavior_ (not just raw values). For state management in JS land, that "service" may simply be a map that holds the keys/values and that's it. In C# land I actually see us storing the `IStateManager` instances themselves in there because we likely wouldn't want to implement them as singleton and then each `IStateManager` would have it's own internal dictionary to hold the keys/values. 

Additionally, in .NET land, one of the things I expect the service collection to do (it doesn't yet) is enumerate all the services that ended up registered in the collection, determine which as `IDisposable` and dispose of them at the end of the turn. I believe this is something that would be expected from a lifetime management perspective.

Finally, I want to go on record one last time and say that if we had whole heartedly adopted dependency injection from the ground up I wouldn't suggest building it this way at all. Instead, everything would be registered in a container, with respective lifetimes and then injected into the middleware/bot as requested. The ship _seems to_ have sailed on that for now, but I hope a vNext (maybe even minor if we don't box ourselves out of it with current design) will consider it strongly.

**Comment by billba** [comment]
@drub0y I think I get this. Do you still agree that this collection is not bot-developer-readable? In other words, no one except a service developer should be accessing it?

**Comment by drub0y** [comment]
@billba Yes, 100% agree. I do not expect developers working at the "bot level" to actually interact directly with the services API itself. I would expect that if the bot itself is meant to work with anything that happens to be stored in there it would be done via a richer API abstraction that hides the fact that it's even using the "services" collection under the covers. 

In fact, I would go so far as to say that this should be a core design goal and that we need to establish these patterns and document them so that developers who are looking to extend the framework with their own middleware/services have guidelines to follow to provide a consistent development experience for bots.

**Comment by billba** [comment]
Okay, I agree with all of that. That leads me to being a fan of `context.services` - it's not only fairly descriptive, but the naming would tend to nudge developers away from just randomly adding stuff to it.

**Comment by drub0y** [comment]
I've rebased on top of all of @cleemullins recent changes that were merged to master and renamed the things that were still `BotContext` to match `TurnContext` now. PR #342 awaiting formal review.

**Comment by tomlm** [comment]
I have approved Drub0y's change which I guess means I approve the DCR :)

**Comment by drub0y** [comment]
Yep, we are done here, thanks @tomlm!
